Port "Manage teams" related dialogs

ArmandoGraterol opened the pull request:
Added the add and edit teams dialogs. I'm excluding the manage whitelist dialog because that would make this PR much bigger than it already is.

![FireShot Capture 793 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214441543-e2b6d6f7-820e-42dc-a035-f4bfa0ae1d46.png)
![FireShot Capture 794 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214441672-a73bacb6-0e4c-47b8-8ddb-a80de7326f28.png)
![FireShot Capture 795 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214441732-fa7b5b89-8048-49df-9b22-a54e291d1929.png)


Contributes to #58

rdig commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialogForm.tsx:
```diff
@@ -0,0 +1,240 @@
+import styles from './CreateDomainDialogForm.css';
+
+const displayName =
+  'common.ColonyHome.CreateDomainDialog.CreateDomainDialogForm';
```
Shouldn't this be under `Dialogs` rather than under `ColonyHome` ?

rdig commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,145 @@
+  Partial<WizardDialogType<object>> &
+  CustomWizardDialogProps;
+
+const displayName = 'common.ColonyHome.EditDomainDialog';
```
Same here

rdig commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialogForm.tsx:
```diff
@@ -0,0 +1,358 @@
+
+import styles from './EditDomainDialogForm.css';
+
+const displayName = 'common.ColonyHome.EditDomainDialog.EditDomainDialogForm';
```
Same here

rdig commented on src/components/common/Dialogs/ManageDomainsDialog/ManageDomainsDialog.tsx:
```diff
@@ -0,0 +1,144 @@
+import { getAllUserRoles } from '~redux/transformers';
+import { canArchitect, hasRoot } from '~utils/checks';
+
+const displayName = 'common.ColonyHome.ManageDomainsDialog';
```
yet another component name that needs to be fixed

rdig reviewed (approved):
Other than components name, all's good here

willm30 commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,118 @@
+  return (
+    <Form<FormValues>
+      defaultValues={{
+        forceAction: false,
```
Does `forceAction` need to be included in the form's payload? It's only used to determine which saga to fire, I believe?

In which case, we could pass the `ForceToggle` component the setState function directly, meaning we wouldn't need to derive the `forced` state from the form state, which would mean we could remove lines 98-102

ArmandoGraterol replied:
It doesn't need to be included in the payload, but it's also not only used to determine which saga to fire as we have various conditionals inside the dialog forms that use this value, mainly revolving around permission stuff. Do we believe that in this case is it worth it to prop drill `setForceAction` and `forceAction`? 

FYI in case we prop drill those values instead of using the form context, currently, it would travel like this in almost every dialog that has an action that can be made into a motion:
```
Dialog > DialogForm > DialogHeading > ForceToggle
Dialog > DialogForm > DialogHeading > MotionDomainSelect (only forceAction)
Dialog > DialogForm > useDialogActionPermissions (only forceAction)
Dialog > DialogForm > DialogControls (only forceAction)
```

willm30 replied:
Ok, let's leave as it is then. If it were just prop drilling to the ForceToggle component then I might've thought it was worth it but passing to multiple components probably isn't preferable to what we have currently. I was trying to think of a way to make the forced state accessible to the form's contents and the form itself. The only way I can think of would be to wrap the dialog in a context that stores the forced state, but that's just me thinking out loud (have not tested). Also that would be a separate issue if it did work. But mentioning just so you know what I was thinking.
![dialog](https://user-images.githubusercontent.com/64402732/219331575-b120a854-1918-48a4-b9ea-576d021189bf.png)

ArmandoGraterol replied:
Yeah, I was also thinking of ways of including this state in a more general component like the `ActionHookForm`, but I haven't encountered a good, easy way of doing it. Thank you for your thoughts! I will think about this some more

willm30 commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,118 @@
+  Partial<WizardDialogType<object>> &
+  CustomWizardDialogProps;
+
+const displayName = 'common.EditDomainDialog';
```
I can't remember from a previous review if these displayNames are supposed to include `common.Dialogs...`?

ArmandoGraterol replied:
I don't think so, that would mean that there's a `Dialogs` component when that's just a containing folder

willm30 commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialogForm.tsx:
```diff
@@ -0,0 +1,156 @@
+      )}
+      <DialogSection>
+        <div className={styles.nameAndColorContainer}>
+          <div className={styles.domainName}>
```
I see that this combination of domain and color select is also included in the EditDomainDialogForm, so I think could be moved to own component.

willm30 commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialog.tsx:
```diff
@@ -0,0 +1,97 @@
+        annotationMessage: '',
+        // motionDomainId: Id.RootDomain,
+      }}
+      submit={getFormAction('SUBMIT')}
```
This `getFormAction` logic is being duplicated in every form. 

An alternative is to move the part concerned with the action type suffix into a util, which gets called directly inside the ActionForm

``` js
export const getFormAction = (
  action: ActionTypes,
  actionType: 'ERROR' | 'SUCCESS',
) => `${action}_${actionType}`;
```

then 

``` js
const ActionHookForm = <V extends Record<string, any>>({
  actionType,
  onSuccess,
  onError,
  onSubmitError,
  transform,
  ...props
}: Props<V>) => {
  const submit = actionType;
  const error = getFormAction(actionType, 'ERROR');
  const success = getFormAction(actionType, 'SUCCESS');
 ...
 }
```

Then you only need to pass in the actionType, the logic for which can be kept in the parent component. E.g.

``` tsx 
const actionType = !isForce /*&& isVotingExtensionEnabled*/
      ? ActionTypes.MOTION_DOMAIN_CREATE_EDIT
      : ActionTypes.ACTION_DOMAIN_CREATE

```

willm30 replied:
@ArmandoGraterol Adding to this for completeness, Raul mentioned in one of my prs that this should also include a way to pass in the states individually, but that they have default values assigned

https://github.com/JoinColony/colonyCDapp/pull/261#discussion_r1107359898

ArmandoGraterol replied:
Thank you for the update :+1:

willm30 reviewed (commented):
Nice job @ArmandoGraterol!

When I open the Edit Team dialog from a colony I created manually (as opposed to via the script), I'm seeing the following: 

![Screenshot (659)](https://user-images.githubusercontent.com/64402732/219014747-c8b1dded-4370-4ea5-b42c-5e941cd8d358.png)

Which in itself is fine, but drew my attention to the select team dropdown, which presumably should also be greyed out? Also the pointers are inconsistent in the disabled inputs: some show the disabled pointer, some don't. And finally, although I couldn't catch it in the screenshot, the tooltip "?" icon has a title, which overlaps the actual tooltip text and should probably be removed.

Just commenting for now as I left some over comments, but good stuff 🙂

rdig reviewed (approved):
Will has been very thorough in his review here, so I don't really have anything that I can point to.

I'm pre-approving this, but please address Will's issues, especially the ones that deal with extracting into separate components as that will help with readability and code reuse

ArmandoGraterol commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialogForm.tsx:
```diff
@@ -0,0 +1,258 @@
+  //   [setFieldValue],
+  // );
+
+  useEffect(() => {
```
Hmmm, I thought I removed this old code, perhaps I did it in another place and forgot this one :thinking: Thank you!

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/validation.ts:
```diff
@@ -21,7 +21,7 @@ const MSG = defineMessages({
 const validationSchema = object()
   .shape({
-    domainId: string().required(),
+    fromDomain: number().required(),
```
I'll wait for a conclusion here to update the code if we decide to store the id as a number or string but yeah I think initially the domain ids are coming as numbers  (For example, `Id.RootDomain` is a number).

jakubcolony replied:
Good spot @willm30, that was actually an oversight on my side, #263 changes the domain fields to `Int`

jakubcolony commented on src/utils/domains.ts:
```diff
@@ -1,10 +1,21 @@
+  return domains
+    .filter(notNull)
+    .find(({ id }) =>
+      skipRootDomain && (domainId === undefined || domainId === Id.RootDomain)
+        ? Number(id) !== Id.RootDomain
+        : Number(id) === domainId,
```
This logic is actually wrong - it should be comparing the `domainId` against `nativeId`.

jakubcolony reviewed (changes requested):
Nice work @ArmandoGraterol!

I'm requesting changes mainly due to the issue with missing permissions UI:
I'm experiencing a similar issue to Will, but it's with the colonies created via the script. I see a message regarding missing permission but the team selector stays active and when I select one, the entire form becomes editable:
<img width="523" alt="image" src="https://user-images.githubusercontent.com/112586815/219483554-90a93001-c58b-40a8-8856-50d49302d5da.png">
<img width="536" alt="image" src="https://user-images.githubusercontent.com/112586815/219483654-67a37787-0104-43c7-8176-78a8053aa899.png">

Just a small note, in other forms we're able to show a required error that doesn't include the camelCased field name - could that also be done here?
<img width="484" alt="image" src="https://user-images.githubusercontent.com/112586815/219481660-4d15608c-31ea-47e5-998d-5c0f89a4099c.png">

The `useEnabledExtensions` should be now available in `port/colony-actions` so there's no need to comment it out anymore

ArmandoGraterol commented:
@jakubcolony 
> The useEnabledExtensions should be now available in port/colony-actions so there's no need to comment it out anymore

I see that this new hook only provides whether the extension is enabled or not, but the old one also returned the version of the voting reputation extension. Would you know if that's obtained now from another place? Or do we want to update this new hook to also return that value?

jakubcolony commented:
@ArmandoGraterol

willm30 commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialogForm.tsx:
```diff
@@ -0,0 +1,123 @@
+      </DialogSection>
+      {!userHasPermission && (
+        <DialogSection>
+          <PermissionRequiredInfo requiredRoles={[ColonyRole.Architecture]} />
```
The `requiredRoles` const can be reused here

willm30 commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialogForm.tsx:
```diff
@@ -0,0 +1,123 @@
+      {!userHasPermission && (
+        <DialogSection appearance={{ theme: 'sidePadding' }}>
+          <NoPermissionMessage
+            requiredPermissions={[ColonyRole.Architecture]}
```
And here

willm30 commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialogForm.tsx:
```diff
@@ -0,0 +1,123 @@
+
+  const inputDisabled = !userHasPermission || onlyForceAction || isSubmitting;
+
+  const cannotCreateMotion =
```
Could the logic above the view here be wrapped up into a hook that only exposes the data needed by the view? (eg `useCreateDomainDialogForm`)

willm30 commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -221,46 +181,33 @@ Props) => {
-          </div>
+          <NoPermissionMessage
+            requiredPermissions={[
+              ColonyRole.Administration,
```
There's a  required roles const above

willm30 commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -137,17 +98,16 @@ Props) => {
-  const canMakePayment = userHasPermission; // && isOneTxPaymentExtensionEnabled;
+  const canMakePayment = userHasPermission && isOneTxPaymentEnabled;
 
   const inputDisabled = !canMakePayment || onlyForceAction || isSubmitting;
```
Similar thought about wrapping up as much logic as possible so we're only exposing what the view needs (and reducing the number of lines someone needs to read through before they get to the view itself)

willm30 commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -221,46 +181,33 @@ Props) => {
           disabled={
-            // cannotCreateMotion ||
-            !isValid || !!customAmountError || inputDisabled
+            cannotCreateMotion ||
```
I think this should be moved to a const (e.g `areControlsDisabled`)

willm30 commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,102 @@
+const validationSchema = object()
+  .shape({
+    forceAction: boolean().defined(),
+    domainName: string().max(20).required(),
```
Can these have the appropriate messages to avoid the following?

![domainname](https://user-images.githubusercontent.com/64402732/224707511-ce199e5e-3ec1-405c-bf83-915097e56121.png)

willm30 commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialogForm.tsx:
```diff
@@ -0,0 +1,195 @@
+  colony,
+  colony: { domains },
+}: ActionDialogProps) => {
+  const {
```
Same thought here re wrapping this logic into a hook (applies to all components really where there's more than a few lines of business logic above the view (so I don't repeat myself any more in the comments))

willm30 commented on src/components/common/Dialogs/MintTokenDialog/MintTokenDialogForm.tsx:
```diff
@@ -141,41 +131,19 @@ const MintTokenDialogForm = ({ colony, back }: ActionDialogProps) => {
-              }}
-            />
-          </div>
+          <NoPermissionMessage requiredPermissions={[ColonyRole.Root]} />
```
required roles const above

willm30 reviewed (commented):
UI working for me 🙂 Left some comments on tidying up some of the logic in the components

jakubcolony reviewed (commented):
All seems to be working well, great job! 👍 
I left a few comments about some minor improvements

ArmandoGraterol commented on src/hooks/useEnabledExtensions.tsx:
```diff
@@ -1,9 +1,11 @@
 
-const useEnabledExtensions = () => {
-  const { installedExtensionsData, loading } = useExtensionsData();
+const useEnabledExtensions = (colony: Colony | undefined) => {
```
If we add `?`, this would mean that the param is optional no? We could call `useEnabledExtensions()` if that was the case, which is wrong because without a `colony` the hook will never return the data, as opposed to requiring a `colony` param that can be defined or not (We are assuming that eventually the `colony` will be defined, if not already).

jakubcolony replied:
Fair enough, thanks 👍

ArmandoGraterol commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialog.tsx:
```diff
@@ -0,0 +1,100 @@
+  const [isForce, setIsForce] = useState(false);
+  const navigate = useNavigate();
+
+  const { isVotingReputationEnabled } = useEnabledExtensions(colony);
```
It's possible because we could use that hook in the `NewActionButton` and pass the data from there to the dialog wizard, however, we have to take into consideration that we also need to pass `isOneTxPaymentEnabled` and `votingReputationVersion`. So, we would need to pass 3 new props or 1 with the 3 props inside it (something like `extensionData`?).

What do you think?

jakubcolony replied:
I don't have a preference, whatever feels right for you. You could possibly introduce a type that we'd use for both the return value of `useEnabledExtensions` and the 1 props you mentioned

ArmandoGraterol commented on src/components/common/NewActionButton/NewActionButton.tsx:
```diff
@@ -82,7 +79,7 @@ const NewActionButton = ({ filteredDomainId }: Props) => {
   const hasRegisteredProfile = !!user?.name && !!user.walletAddress;
   // const mustUpgrade = colonyMustBeUpgraded(colony, networkVersion as string);
-  // const isLoadingData = isLoadingExtensions || isLoadingUser;
+  const isLoadingData = loading; // || isLoadingUser;
```
This is a good question, thank you! I have noticed that there's a `walletConnecting` available on `useAppContext` that should be the equivalent of the `isLoadingUser` so I'm gonna implement that

ArmandoGraterol commented on src/components/shared/ColorSelect/ColorSelect.tsx:
```diff
@@ -57,20 +57,17 @@ const ColorSelect = ({
-    );
-    return [
-      ...colors.map((color) => {
+  const options = useMemo<ComponentProps<typeof Select>['options']>(
```
Nice one! Thank you

ArmandoGraterol commented on src/components/shared/DomainFundSelectorSection/helpers.ts:
```diff
@@ -1,11 +1,18 @@
-export const getDomainOptions = (colonyDomains: Domain[]) =>
+export const getDomainOptions = (
+  colonyDomains: Domain[],
+  skipRootDomain?: boolean,
```
Sounds good to me :ok_hand:  I'll make the changes

ArmandoGraterol commented:
@willm30 I've added a hook called `useActionDialogStatus` in which I tried to add as much logic in common between the dialog forms as I could, let me know what you think!

@jakubcolony I've reverted the changes to `useEnabledExtensions` and added the data as props to the action dialogs.

jakubcolony commented on src/components/common/Dialogs/CreatePaymentDialog/helpers.ts:
```diff
@@ -32,3 +36,33 @@ export const getCreatePaymentDialogPayload = (colony: Colony, payload: any) => {
+  enabledExtensionData: EnabledExtensionData,
+) => {
+  const { getValues } = useFormContext();
+  const values = getValues();
```
So apparently, after talking to @willm30, I think you should use the `watch` function here. `getValues` doesn't guarantee that the component will get re-rendered if `values.fromDomain` changes
On a side note, you could destructure `fromDomain`

jakubcolony commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,124 @@
+import Dialog, { DialogProps, ActionDialogProps } from '~shared/Dialog';
+import { ActionHookForm as Form } from '~shared/Fields';
+
+import { Domain } from '~types';
```
This line gives me a TS unused variable error

jakubcolony commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,124 @@
+import { getDomainOptions } from '~shared/DomainFundSelectorSection/helpers';
+
+import EditDomainDialogForm from './EditDomainDialogForm';
+import { getEditDomainDialogPayload, notRootDomain } from './helpers';
```
As above

ArmandoGraterol replied:
Do you mean it's appearing as if they are unused? It's not the case for me :thinking:

jakubcolony commented on src/components/shared/ColorSelect/ColorSelect.tsx:
```diff
@@ -36,11 +38,13 @@ const displayName = 'ColorSelect';
   name = 'color',
 }: Props) => {
+  const { getValues } = useFormContext();
+  const activeOption = getValues(name);
```
I think this should also be using `watch`

jakubcolony commented on src/components/shared/Fields/Select/HookForm/HookFormSelect.tsx:
```diff
@@ -53,11 +53,12 @@ const HookFormSelect = ({
     setValue,
   } = useFormContext();
-  const error = errors[name] as Message;
+  const error = errors[name]?.message as Message;
```
Should be `as Message | undefined`

jakubcolony commented on src/components/shared/TokenAmountInput/TokenAmountInput.tsx:
```diff
@@ -86,7 +89,7 @@ const TokenAmountInput = ({ colony, disabled }: Props) => {
             <EthUsd
               // appearance={{ theme: 'grey' }}
-              value={values.amount}
+              value={values.amount.replace(/,/g, '') || 0} // @NOTE: Get the amount without commas as BigNumber doesn't like them
```
I think the problem should be fixed at the root (`FormattedInputComponent` should be storing the Cleave's raw value instead of formatted value which may contain the comma separators). I fixed it in #320 so do you want to add a TODO comment here to change it later?

ArmandoGraterol replied:
For sure, I will add it to the other instances where we are removing commas like this as this ain't the only instance.

jakubcolony reviewed (changes requested):
The UI works mint, but I had a few comments on the new changes. Otherwise nice work!

ArmandoGraterol commented on src/components/common/Dialogs/EditDomainDialog/EditDomainDialog.tsx:
```diff
@@ -0,0 +1,124 @@
+  enabledExtensionData,
+}: Props) => {
+  const colonyDomains =
+    colony.domains?.items.filter(
+      (domain) => notNull(domain) && notRootDomain(domain),
+    ) || [];
```
I also considered this and seeing that you concur with what I thought then will do :+1:

willm30 commented on src/components/shared/TokenEditDialog/TokenEditDialog.tsx:
Belongs in `common/Dialogs`?

willm30 commented on src/components/shared/TokenAmountInput/TokenAmountInput.tsx:
This probably also belongs in the dialogs folder since it appears to be an abstraction specific to a couple of dialogs only

willm30 commented on src/components/shared/DomainFundSelectorSection/DomainBalance.tsx:
Same thought for this one re moving somewhere inside `common/Dialogs`

willm30 reviewed (approved):
Nice one Armando. I think some of the components either introduced here or edited here actually belong inside a `common/Dialogs` as opposed to `shared`, since they are only and may only be expected to be used in some of the dialogs. 

Also, I think this pr would be easier to review if it were even smaller, e.g. one pr per dialog introduced. (That's definitely something I want to get better at also, having prs address the smallest logical / practical set of changes possible).  

But don't let me hold you up 🙂 UI's good and all will be revisited again in the not too distant future as part of the ui refresh.

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/validation.ts:
```diff
@@ -39,7 +39,7 @@ const validationSchema = object()
         () => MSG.amountZero,
         (value) => {
-          const numberWithoutCommas = (value || '0').replace(/,/g, '');
+          const numberWithoutCommas = (value || '0').replace(/,/g, ''); // @TODO: Remove this once the fix for FormattedInputComponent value is introduced.
```
Do you mean the downgrading of `cleave.js`? Nope, this is related to this fix that @jakubcolony implemented: https://github.com/JoinColony/colonyCDapp/pull/320/commits/dd6f97d886bb3207d2c29018a6ebdd85544e86ae

Basically, the cleave was using the formatted value of the input instead of the raw value like in the Dapp. So, we would get errors when dealing with the value stored by the form because they would have commas `,`. This is a solution I implemented before Jakub informed me of the issue with the cleave value.

I should remove this line when Jakub's fix makes its way into the branch.

ArmandoGraterol commented on src/components/shared/CannotCreateMotionMessage/CannotCreateMotionMessage.tsx:
Sure, sounds good to me

jakubcolony reviewed (approved):
All good, nice work! 👍

willm30 commented on src/components/common/Dialogs/CreateDomainDialog/CreateDomainDialog.tsx:
```diff
@@ -0,0 +1,102 @@
+        onSuccess={close}
+      >
+        {({ watch }) => {
+          const forceActionValue = watch('forceAction');
+          if (forceActionValue !== isForce) {
+            setIsForce(forceActionValue);
+          }
```
The way things are currently set up, it's needed in the `actionType` variable which lives outside the form. I did mention in another pr we could look into storing the dialog state in the dialog provider, but that would need looking into separately.
